[ADC] Add multipoint correction and/or binning to ADC

TD-er opened the pull request:
The ADC plugin (for internal ADC) could only apply a 2-point calibration.
But this makes it a bit hard to correct for non-linear corrections like when using a fixed resistor in series with a variable resistor.
This can be corrected using the formula, but this only is applied on the output value.
If based on a single sample, this can be rather noisy and based on oversampled data, it may also yield unexpected results.

For example when using a sensor which has a discrete number of resistor values like some level sensors, or a wind vane in a weather station, using oversampling may make the output value even more erratic.
Especially when such a sensor may double the number of output steps by connecting 2 fixed resistors in parallel.
This would result in the measured resistance to jump around between neighboring measuring steps.

A typical use case is the wind vane of a very popular set, the DS-15901.
![image](https://user-images.githubusercontent.com/3751318/170365605-b6773629-63db-4f14-a658-83a2ce0dfc4a.png)
In this example, the computed measured voltages are when using a 10k resistor in series with the wind vane, connected to 3V3 and GND.
![image](https://user-images.githubusercontent.com/3751318/170358015-cacb4909-90b8-47cc-9bfe-675945aa8793.png)

The charts show the relation between measured voltage and wind direction in degrees.

For this, the ADC values should **not** be averaged, but rather counted in a number of bins. (16 in this use case)

This needs to be processed using the following steps in this specific order:
- 2-point calibration to convert the ADC values into a true voltage (e.g. 0 -> 0V & 4095 -> 3.3V on an ESP32)
- Formula to compute the resistance from the measured voltage `(-10000*%value%)/(%value%-3.3)` based on a 10k resistor and 3V3.
- Pick the correct bin, based on the computed resistance, using the resistance values from the datasheet.
- Output the associated value (wind direction in degree in this use case) from the bin with the highest count. (bin center value defined by the computed resistance)

![image](https://user-images.githubusercontent.com/3751318/170371279-30449de4-a5e0-48a2-aef6-2c0f9d95f878.png)
Relation of resistance per bin.

This way, the ADC is sampled with the same higher sampling rate, comparable with oversampling.
But since each sample is processed in each of the above mentioned steps, one can still get a stable reading even though some measured voltages are quite close to each other.
This would be otherwise impossible to do in rules as it needs very complex rules and those rules cannot be processed at roughly 10 samples/sec.

TD-er committed:
[P002 ADC] Move PluginTaskData to separate file

TD-er committed:
[ADC] Add binning and multi-point processing of values

TD-er committed:
[ADC] Improve binning stability

TD-er committed:
[ADC] Speed up formula parsing

TD-er committed:
[ADC] Add caching for binning range

TD-er committed:
[ADC] Add fields to load/save multipoint values.

TD-er committed:
[ADC] Add option to paste many variables at once

TD-er commented:
Another featured added here is to be able to paste a number of values at once as you otherwise need to enter 32 values to get a set of 16 multpoint processing values.

Example of the wind vane use case:
![image](https://user-images.githubusercontent.com/3751318/170362208-2bd7c8cd-94f6-40d6-b9c4-a06be960472d.png)

This can be entered at once using some pre-formatted text like this, taken literally from a spreadsheet (thus values separated by a tab and CRLF)
```
33000	0
6570	22.5
8200	45
891	67.5
1000	90
688	112.5
2200	135
1410	157.5
3900	180
3140	202.5
16000	225
14120	247.5
120000	270
42120	292.5
64900	315
21880	337.5
```

To enter this set all in one go, the checkbox for "Split-paste Multipoint Fields" must be checked and then the values must be paste in just any of the fields for Point 1...N
N.B. This will also clear the unused fields.

After submit, the page will reload and the values will be sorted on the input value.

TD-er committed:
Merge branch 'mega' into feature/ADC_multipoint_calib

TD-er committed:
[Cleanup] Small optimizations to reduce bin size

TD-er committed:
[Cleanup] Reduce loading controller settings when not ready to process

TD-er committed:
[Cleanup] Add function addFormSelector_YesNo to remove code duplication

TD-er committed:
[Cleanup] Reduce build size by combining controller queue types

TD-er committed:
[Controller Queue] Add missing includes

TD-er committed:
[PVS Studio] Fix some hints provided by PVS Studio

TD-er committed:
[P073, PVS Studio] Fix bug in font handling for Siekoo display

Found by PVS Studio

TD-er committed:
[ADC] Fix compiler warnings + remove binning for LIMIT_BUILD_SIZE builds

TD-er committed:
Merge branch 'mega' into feature/ADC_multipoint_calib

TD-er committed:
Merge branch 'mega' into feature/ADC_multipoint_calib

TD-er committed:
Merge branch 'mega' into feature/ADC_multipoint_calib

TD-er committed:
Merge branch 'mega' into feature/ADC_multipoint_calib

TD-er committed:
Merge branch 'mega' into feature/ADC_multipoint_calib

TD-er committed:
Merge branch 'mega' into feature/ADC_multipoint_calib

TD-er committed:
[ADC] Add support for ESP32 ADC factory calibration

TD-er committed:
[PlatformIO] Fix boot loop issues on 16M ESP32 boards

TD-er committed:
[ADC] Fix building on ESP32-S2

TD-er committed:
[Cleanup] Uncrustify Helpers/Hardware.cpp

TD-er committed:
[ADC] Fix build on ESP8266

TD-er committed:
[ADC] Fix crash when importing older ADC task settings enabling binning

TD-er committed:
[ADC] Add charts to show the factory calibration

TD-er committed:
[ADC] Hide Factory calibration data + fix build on ESP8266

TD-er committed:
[ADC] Add chart axis options (e.g. label)

TD-er committed:
[ADC] Add chart to show simulated ADC to output

TD-er committed:
[ADC] Add attenuation option for ESP32

tonhuisman commented on src/src/PluginStructs/P002_data_struct.cpp:
```diff
@@ -69,9 +81,30 @@ void P002_data_struct::load(struct EventStruct *event)
+  const float currentValue = P002_data_struct::getCurrentValue(event, raw_value);
+
 # if defined(ESP32)
   addHtml(F("<TR><TD>Analog Pin:<TD>"));
```
Should this be `addRowLabel(F("Analog Pin"));` ?

TD-er replied:
Probably yes...
One of the very few old lines of code I have not changed ;)

TD-er committed:
[Cleanup] Remove some unneeded HTML code

TD-er committed:
[ADC] Fix building on builds with LIMIT_BUILD_SIZE

TD-er committed:
[ADC] Add pseudo task values via PLUGIN_GET_CONFIG

TD-er committed:
[ADC] Fix reading + cleanup UI

TD-er committed:
[ADC] Add PluginStats to keep history of values

TD-er committed:
[ADC] Fix build issue on LIMIT_BUILD_SIZE

TD-er committed:
[PluginStats] Add generic statistics for tasks

TD-er committed:
[Plugin Stats] Add option for average over last N samples.

TD-er committed:
[Plugin Stats] Use task specific error value

TD-er committed:
[Plugin Stats] Make Plugin Stats optional per task value

TD-er committed:
[Plugin Stats] Add version transitions to ExtraTaskSettings

TD-er committed:
[Plugin Stats] Enable Plugin Stats for all plugins with task data

... and outputting float values

TD-er committed:
[Plugin Stats] Show statistics + historic samples on all supported tasks

TD-er committed:
[Plugin Stats] Add missing plugins to use the plugin stats too

TD-er committed:
[PluginStats] Add build flags USES_PLUGIN_STATS and USES_CHART_JS

TD-er committed:
Merge branch 'mega' into feature/ADC_multipoint_calib

TD-er committed:
[PluginStats] Exclude PluginStats and ChartJS from NeopixelESP8266 build

TD-er committed:
[ADC] Change order of items shown in statistics

tonhuisman commented on src/src/PluginStructs/P002_data_struct.cpp:
```diff
@@ -243,38 +243,36 @@ void P002_data_struct::webformLoad(struct EventStruct *event)
-        formatADC_statistics(F("ADC Peak High"), _plugin_stats[0]->getPeakHigh());
-      }
+    if (_plugin_stats[0]->getNrSamples() > 0) {
+      addRowLabel(F("Avg. ouput value"));
```
Typo?

TD-er replied:
Where?

tonhuisman replied:
`"Avg. ouput value"` -> `"Avg. output value"`

TD-er replied:
fixed

TD-er committed:
[ADC] Use set nr of decimals for stats

TD-er committed:
[Cleanup] Move PluginTaskData_base structs to Pxxx_data_struct.h/.cpp

TD-er committed:
[SVG] Revert changes to draw SVG since we're now using ChartJS

TD-er committed:
[ADC] Add computed output values for peaks + fix typo

TD-er committed:
[ADC] Document new options

TD-er committed:
[Plugin Stats] Add stats to plugins to support stats.

TD-er committed:
[Docs] Document accessing Plugin Stats

TD-er committed:
[Docs] Document typical task setup page and task value stats

TD-er committed:
[Plugin Stats] Enable Stats for Sysinfo tasks

TD-er committed:
[Plugin Stats] Reduce memory usage when no stats are used.

TD-er committed:
[Build] Do not limit build size for custom builds with VCC. (#4129)

Fixes: #4129

TD-er committed:
[Plugin Stats] Make labels more clear.

TD-er committed:
[Plugin Stats] Fix strange issues PLUGIN_INIT 

There was a different code path for PLUGIN_INIT_ALL compared to starting a task via PLUGIN_INIT after boot.

TD-er committed:
[Plugin Stats] Fix editing disabled tasks

TD-er committed:
[Plugin Stats] Fix memory leak / low loop count

TD-er committed:
[ESP32] Add cache for ExtraTaskSettings due to load speed on ESP32
